### PR TITLE
Experimental: modified version of Kambui Nurse's spectral package

### PR DESCRIPTION
**Pull request by shykes**

### CODE REVIEW

**Comment by shykes** [comment]
cc @Siafu @jpadams

**Comment by jpadams** [comment]
@shykes made PR against your repo with lint fix. Let me know if that element is necessary.

File: pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	// Human-readable text output
+	output: strings.Join([
+		for _, m in logs {
+			m.message
+		},
+	], "\n")
+
+	// Location of JSON/YAML documents.
+	// Can be either a file, a glob or fetchable resource(s) on the web.
+	// Example: ["*.json", "*.yaml", "https://github.com/dagger/dagger/docker-compose.yaml"]
```
Comment (shykes) [inline pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue:48]: note @siafu I was wrong, spectral supports globs natively, no shell expansion necessary

File: pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	container: docker.#Run & {
+		// Default container image can be overrided.
+		//   'spectral' must be installed and in the PATH.
+		input:              docker.#Image | *_buildDefaultImage.output
+		_buildDefaultImage: docker.#Pull & {
+			source: "stoplight/spectral"
+		}
+		// Ugly hack to override the entrypoint from the default image
+		//  FIXME: how do we tuck this into the scope of the default image?
+		entrypoint: []
```
Comment (shykes) [inline pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue:114]: Any ideas @helderco ?

Comment (helderco) [reply pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue:114]: You can use `docker.#Set` to override the default image and use the output of that.

File: pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+		// increase verbosity
+		"--verbose"?: bool
+
+		// no logging - output only
+		"--quiet"?: bool
+	}
+
+	// Setup a Docker container to run the spectral CLI
+	container: docker.#Run & {
```
Comment (TomChv) [inline pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue:105]: Why you didn't not directly embedded `docker.#Run`? That would avoid using a for loop to set all `flags`

You can also use disjunction to handle `ruleset` usage

**Review by TomChv — COMMENTED** [review]
That's a complete package, nice work :rocket:

File: pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue (hunk 3 above)
Comment (shykes) [reply pkg/universe.dagger.io/x/solomon@dagger.io/spectral/lint.cue:105]: Initially it was embedded. But:

* When at some point we added an `input` and `output` field, it conflicted with the same field names in `docker.#Run`
* When moving all docker-specific fields in the `docker.#Run & { }`, there were scoping issues: non-docker fields now have to re-declare docker field to reference them. It's cleaner to have an explicit `container` key to reference.
* Same pattern as `bash.#Run`, for consistency

**Comment by shykes** [comment]
I'm stuck on a cue lint error... I've updated my branch with `cue fmt -s` everywhere (forgot to lint the tests...) but still getting an error.

* I can't reproduce the error locally with `dagger do lint cue` (does anyone use this?)
* The CI job here offers no detail of the actual error...

So I'm stuck. Any help would be appreciated :)

**Comment by shykes** [comment]
Spare an approval @jpadams  @helderco ? :)

**Review by jpadams — APPROVED** [review]
LGTM!

**Comment by shykes** [comment]
Thanks all!

**Comment by jpadams** [comment]
out

**Pull request merged**